packages/cli/lib/LocalDevManager.js

```diff
@@ -217,10 +251,54 @@ class LocalDevManager {
     }
   }
 
+  startWatching(runnableComponents) {
+    this.watcher = chokidar.watch(this.projectDir, {
+      ignoreInitial: true,
+    });
+
+    const configPaths = runnableComponents
+      .filter(({ type }) => type === COMPONENT_TYPES.app)
+      .map(component => {
+        const appConfigPath = path.join(component.path, APP_COMPONENT_CONFIG);
+        return appConfigPath;
+      });
+
+    const projectConfigPath = path.join(this.projectDir, PROJECT_CONFIG_FILE);
+    configPaths.push(projectConfigPath);
+
+    this.watcher.on('add', filePath => {
+      this.handleWatchEvent(filePath, WATCH_EVENTS.add, configPaths);
+    });
+    this.watcher.on('change', filePath => {
+      this.handleWatchEvent(filePath, WATCH_EVENTS.change, configPaths);
+    });
+    this.watcher.on('unlink', filePath => {
+      this.handleWatchEvent(filePath, WATCH_EVENTS.unlink, configPaths);
+    });
+    this.watcher.on('unlinkDir', filePath => {
+      this.handleWatchEvent(filePath, WATCH_EVENTS.unlinkDir, configPaths);
+    });
+  }
+
+  async stopWatching() {
+    await this.watcher.close();
+  }
+
+  handleWatchEvent(filePath, event, configPaths) {
+    if (configPaths.includes(filePath)) {
+      this.logUploadWarning(
+        i18n(`${i18nKey}.uploadWarning.configEdit`, {
+          path: path.relative(this.projectDir, filePath),
+        })
+      );
+    } else {
+      this.devServerFileChange(filePath, event);
```

Comment:
This makes it so the dev servers can make their own checks for config changes within their respective components. If they detect a file change that requires an upload, they can call the `onUploadRequired` callback that gets passed down to them in `setup`.